lib/commands/audit.js

```diff
@@ -1,8 +1,345 @@
 const Arborist = require('@npmcli/arborist')
 const auditReport = require('npm-audit-report')
-const reifyFinish = require('../utils/reify-finish.js')
-const auditError = require('../utils/audit-error.js')
+const chalk = require('chalk')
+const fetch = require('npm-registry-fetch')
+const localeCompare = require('@isaacs/string-locale-compare')('en')
+const npa = require('npm-package-arg')
+const pacote = require('pacote')
+
 const ArboristWorkspaceCmd = require('../arborist-cmd.js')
+const auditError = require('../utils/audit-error.js')
+const {
+  registry: { default: defaultRegistry },
+} = require('../utils/config/definitions.js')
+const log = require('../utils/log-shim.js')
+const pulseTillDone = require('../utils/pulse-till-done.js')
+const reifyFinish = require('../utils/reify-finish.js')
+
+const sortAlphabetically = (a, b) => localeCompare(a.name, b.name)
+
+class VerifySignatures {
+  constructor (tree, filterSet, npm, opts) {
+    this.tree = tree
+    this.filterSet = filterSet
+    this.npm = npm
+    this.opts = opts
+    this.keys = new Map()
+    this.invalid = new Set()
+    this.missing = new Set()
+    this.audited = 0
+    this.verified = 0
+    this.output = []
+    this.exitCode = 0
+  }
+
+  async run () {
+    const start = process.hrtime.bigint()
+
+    // Find all deps in tree
+    this.edges = this.getEdgesOut(this.tree.inventory.values(), this.filterSet)
+    if (this.edges.size === 0) {
+      throw new Error('No dependencies found in current install')
+    }
+
+    // Prefetch and cache public keys from used registries
+    const registries = this.findAllRegistryUrls(this.edges)
+    for (const registry of registries) {
+      const keys = await this.getKeys({ registry })
+      if (keys) {
+        this.keys.set(registry, keys)
+      }
+    }
+
+    await Promise.all([...this.edges].map((edge) => this.getVerifiedInfo(edge)))
```

Comment:
@wraithgar this currently executes all manifest requests concurrently, which seems to work but the spinner is really laggy so I think it's exhausting resources and might cause issues on weaker machines (I'm running with 64gb of ram and 8 cores..).

Do you have a pattern for limiting concurrency in the CLI?

I've used this library in the past with great effect: https://github.com/rxaviers/async-pool. I did a quick test locally and limiting to around ~25 concurrent requests seemed to shave a second of the verification time from my machine on the cli deps and the spinner is smooth.

Reply:
but using ES9 features so probably a no-go for the cli 🤔

Reply:
Node will look into this: https://github.com/supercharge/promise-pool